False positive for missing-yield-type-doc in annotated generator

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?: pip install pylint astroid --pre -U

-->

### Steps to reproduce
1. Enable the `pylint.extensions.docparams` plugin in pylintrc
2. Run against the following file:
```
"""Small repro"""

import typing


def generator() -> typing.Iterator[int]:
    """A simple function for reproducing a pylint bug.

    Yields:
        The number 0
    """
    yield 0
```

### Current behavior
```
pylint repro.py 
************* Module repro
repro.py:6:0: W9014: Missing yield type documentation (missing-yield-type-doc)
```

### Expected behavior
No warning.

### pylint --version output
```
pylint 2.4.2
astroid 2.3.1
Python 3.7.4 (default, Aug 27 2019, 22:43:30) 
[Clang 10.0.0 (clang-1000.11.45.5)]
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -53,6 +53,11 @@ Release date: TBA
 
   Close #3150
 
+* Fixed ``missing-yield-type-doc`` getting incorrectly raised when
+  a generator does not document a yield type but has a type annotation.
+
+  Closes #3185
+
 
 What's New in Pylint 2.4.2?
 ===========================
@@ -386,7 +391,7 @@ must rename the associated identification.
 
 * Allow the choice of f-strings as a valid way of formatting logging strings.
 
-Closes #2395
+  Closes #2395
 
 * Added ``--list-msgs-enabled`` command to list all enabled and disabled messages given the current RC file and command line arguments.
 

--- a/pylint/extensions/docparams.py
+++ b/pylint/extensions/docparams.py
@@ -316,7 +316,7 @@ class DocstringParameterChecker(BaseChecker):
         if not doc_has_yields:
             self.add_message("missing-yield-doc", node=func_node)
 
-        if not doc_has_yields_type:
+        if not (doc_has_yields_type or func_node.returns):
             self.add_message("missing-yield-type-doc", node=func_node)
 
     def visit_yieldfrom(self, node):

--- a/tests/extensions/test_check_return_docs.py
+++ b/tests/extensions/test_check_return_docs.py
@@ -76,21 +76,6 @@ class TestDocstringCheckerReturn(CheckerTestCase):
         ):
             self.checker.visit_return(return_node)
 
-    def test_sphinx_returns_annotations(self):
-        node = astroid.extract_node(
-            '''
-        def my_func(self) -> bool:
-            """This is a docstring.
-
-            :returns: Always False
-            """
-            return False
-        '''
-        )
-        return_node = node.body[0]
-        with self.assertNoMessages():
-            self.checker.visit_return(return_node)
-
     def test_sphinx_missing_return_type_with_annotations(self):
         node = astroid.extract_node(
             '''

--- a/tests/extensions/test_check_yields_docs.py
+++ b/tests/extensions/test_check_yields_docs.py
@@ -445,3 +445,38 @@ class TestDocstringCheckerYield(CheckerTestCase):
         )
         with self.assertAddsMessages(Message(msg_id="redundant-yields-doc", node=node)):
             self.checker.visit_functiondef(node)
+
+    def test_sphinx_missing_yield_type_with_annotations(self):
+        node = astroid.extract_node(
+            '''
+            import typing
+
+            def generator() -> typing.Iterator[int]:
+                """A simple function for checking type hints.
+
+                :returns: The number 0
+                """
+                yield 0
+            '''
+        )
+        yield_node = node.body[0]
+        with self.assertNoMessages():
+            self.checker.visit_yield(yield_node)
+
+    def test_google_missing_yield_type_with_annotations(self):
+        node = astroid.extract_node(
+            '''
+            import typing
+
+            def generator() -> typing.Iterator[int]:
+                """A simple function for checking type hints.
+
+                Yields:
+                    The number 0
+                """
+                yield 0
+            '''
+        )
+        yield_node = node.body[0]
+        with self.assertNoMessages():
+            self.checker.visit_yield(yield_node)

